### ISSUE TITLE
Horde.Supervisor.terminate_child/2 non consistent with DynamicSupervisor 

### ISSUE THREAD

**Issue by xadhoom** [derekkraan/horde#54]
According to Horde docs and function spec, terminate_child/2 should use the child pid as child_id to terminate, like DynamicSupervisor does.

But does not work, and looking into tests seems that we need to use the id of the process (which works), which is (maybe) fine, but should at least be documented.

**Comment by derekkraan** [comment]
Thanks for the bug report. I would gladly accept a pull request to fix this issue. Looking at the documentation, the description is correct but the typespec does specify a `pid`.

**Comment by xadhoom** [comment]
mmmh, not really :)
The doc states that `Works like DynamicSupervisor.terminate_child`, which accepts a pid.

Horde.Supervisor wants an id instead. So I'll be happy to contribute, but I need to understand what is really wanted.

If an `id` is the wanted params, specs & docs must be fixed (glad to do it).
Otherwise, if we want API parity, `supervisor_impl.ex` must be fixed (easy to do, also).
As a third choice, we may support both :) (maybe is the better choice ?)

Let me know what is your idea and I'll try to do something.

**Comment by derekkraan** [comment]
Ah yes, that's true.

Ideally we would have API parity with Elixir's DynamicSupervisor. Right now we don't have that. The main difference is that Horde.Supervisor knows about ids and actually uses them internally, while DynamicSupervisor does not (it ignores child ids).

The more fundamental difference between Horde.Supervisor and DynamicSupervisor is that Horde.Supervisor has a notion of process identity, and DynamicSupervisor does not.

Here's a concrete example:

```
Horde.Supervisor.start_child(supervisor, %{id: "foo", start: {...}})
#=> {:ok, pid}
Horde.Supervisor.start_child(supervisor, %{id: "foo", start: {...}})
#=> {:error, {:already_started, pid}}

DynamicSupervisor.start_child(supervisor, %{id: "foo", start: {...}})
#=> {:ok, pid}
DynamicSupervisor.start_child(supervisor, %{id: "foo", start: {...}})
#=> {:ok, pid}
```

So maybe you're wondering *why* I am using DynamicSupervisor in Horde. The answer is: for the startup and shutdown behaviour. Supervisor starts and shuts down processes in the order in which they are defined. DynamicSupervisor starts and shuts down processes in parallel. So you could say that Supervisor is a more natural choice when its child processes are not alike (but still related to one another) and DynamicSupervisor is a more natural choice when its child processes *are* alike, and otherwise unrelated to one another.

In graph form:

![image](https://user-images.githubusercontent.com/522157/50335301-211f7680-050b-11e9-868e-ffc3c4eb105f.png)

I am reluctant to do this, but I think in this case we might need to abandon the pretense of full api parity with DynamicSupervisor.

Or do you see a way to get closer to DynamicSupervisor while still retaining the unique child_id semantics that we also want? It feels like these features are independent, but somehow still mutually exclusive in the Supervisor and DynamicSupervisor that we have available to us in Elixir.

**Comment by xadhoom** [comment]
Well, in my use case having the unique id is not a problem and I can say is a plus. Also terminating by id is useful (I'm using it for user sessions which are unique by definition, but the same can be said for IoT  and so on). So the DynamicSupervisor approach is very reasonable.

Said that, should be possible to support both semantics, after all in https://github.com/derekkraan/horde/blob/master/lib/horde/supervisor_impl.ex#L80 we can check if `child_id` is a pid and avoid the call to `Process.whereis(Horde.ProcessSupervisor.name(this_node_id, child_id))` (will be needed to get the child_id from pid in this case for state/cdrts updates).

Or (just thinking out loud), why not drop id completely and use pid everywhere instead? After all, if we want to get the process by key, there's the Horde.Registry for that.

**Comment by derekkraan** [comment]
I'll have to think about this one. I'm not sure dropping the `id` completely is what we should want.

In any case, the documentation for `Horde.Supervisor.terminate_child` could definitely be improved.

**Comment by derekkraan** [comment]
I've updated the documentation in 8bf306c. I'm still thinking about the implications of dropping `child_id` altogether.

**Comment by derekkraan** [comment]
Fixed in #122.
